### PR TITLE
Clean up vary logic and ferry all of request_headers through to encoding

### PR DESCRIPTION
**Pull request by capoferro**
**Commit Message:** Clean up vary logic and ferry all of request_headers instead of filtered vary-related headers in preparation for CachePolicy implementation

**Additional Description:** CachePolicyImpl will need all of request headers. While adding that, I cleaned up the vary logic which is the only user of these headers right now.

Risk Level: low
Testing: unit testing
Docs Changes: inline
Release Notes: code changes to pre-alpha stage CacheFilter

### CODE REVIEW

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (dmitri-d) [inline source/extensions/filters/http/cache/cache_headers_utils.cc:283]: Does this need to be a btree_set? Looks like the callers only iterate over results, perhaps a vector would fit better (and cheaper to maintain?)

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (dmitri-d) [inline source/extensions/filters/http/cache/cache_headers_utils.cc:318]: Is this expected? Perhaps an assert is more appropriate here if that should never happen.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (dmitri-d) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:157]: Doesn't vary allow list come from the CacheConfig only? Should that change the listener along with the filter-chain will get re-created. Could you explain what you had in mind here?

**Comment by dmitri-d** [comment]
Could you also add tests for newly added functionality please?

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (capoferro) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:283]: My thought was that it should deduplicate vary entries. It could be a vector if that's not important.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (capoferro) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:157]: This would happen if an entry gets cached, then the config changes and a subsequent request for the same content hits that cache entry, but now the value the entry varied on is no longer permitted. I defaulted to treating it as a cache miss so we'd reinsert from upstream (or choose not to cache, should upstream still provide a vary header value that is not allowed).

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (capoferro) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:318]: I think it is expected. This is supposed to handle cases where the backend provides a value that our config does not permit varying.

**Comment by capoferro** [comment]
> Could you also add tests for newly added functionality please?

My intent was not to add new functionality. If there's something specific you have in mind that is new that I'm not thinking of, I'm happy to add more tests.

The biggest difference is we are no longer filtering the request_headers to just what vary cares about (as future features I'm working on will require that), but that is not new functionality, just passing along more data that is not yet being used.

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (dmitri-d) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:283]: I'm not sure how important it is to de-duplicate header values here. Sgtm if you think it's worth it.

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (dmitri-d) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:318]: Perhaps mention in the comment that the vary list can change and headers cached by the backend can get stale in that case?

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (dmitri-d) [inline source/extensions/filters/http/cache/cache_headers_utils.cc:309]: Should this be a constexpr?

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (dmitri-d) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:157]: I didn't realise `HttpCacheFactory` returns the same instance of `SimpleHttpCache` when `getCache()` is called. Makes sense and please ignore my comment.

File: test/extensions/filters/http/cache/cache_headers_utils_test.cc
Comment (dmitri-d) [inline test/extensions/filters/http/cache/cache_headers_utils_test.cc:624]: Could you add a test for when a vary header isn't in the white-list (unless I missed it)?

**Comment by dmitri-d** [comment]
This generally lgtm; Could you take another look at your tests to check if the cases when a vary header is no longer in the allowed-list have been covered?

**Comment by mattklein123** [comment]
Thanks will wait until @dmitri-d has approved to take a look.

/wait

File: test/extensions/filters/http/cache/cache_headers_utils_test.cc
Comment (capoferro) [reply test/extensions/filters/http/cache/cache_headers_utils_test.cc:624]: Apparently directly linking to a file that's hidden by default doesn't work very well...

The test is this one, L774:
```
TEST(CreateVaryIdentifier, DisallowedHeader)
```

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (dmitri-d) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:309]: The issue here is that the string is dynamically allocated on every invocation of `createVaryIdentifier`; moving it outside of the method and making it a constexpr is an easy win.

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (dmitri-d) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:318]: ping re: comment

**Comment by dmitri-d** [comment]
Tests lgtm, although the CI is unhappy with the coverage (here; https://storage.googleapis.com/envoy-pr/dbf067f/coverage/index.html). `simple_http_cache` coverage (here: https://storage.googleapis.com/envoy-pr/dbf067f/coverage/source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc.gcov.html): could you add a test for `SimpleHttpCache::varyLookup` to cover the case of changed allow vary header list? 
Don't think it's worth testing unimplemented or trivial methods, but @envoyproxy/senior-maintainers have a final say here.

Could you also clean up CodeQL errors?

@mattklein123: this looks good to me, provided above issues have been resolved.

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (capoferro) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:309]: I'm not a C++ expert, so forgive me for being dense.

If I move this out to be a constexpr string_view, then I either have to copy the string_view to a string (and pay the allocation anyway) or I have to not use absl::StrAppend later on, and pay for additional copies there. Is there another way that constexpr string_view saves me allocations that I am not considering?

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (capoferro) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:318]: Done

**Comment by capoferro** [comment]
> could you add a test for SimpleHttpCache::varyLookup to cover the case of changed allow vary header list?

Done. I also fixed a bug in the SimpleHttpCache test that I noticed while in there.

**Comment by mattklein123** [comment]
Please check CI.

/wait

**Comment by capoferro** [comment]
/wait 

still workin on stuff

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/cache_headers_utils.cc:305]: this is pre-existing but can you change these to HeaderSeparator and InValueSeparator per convention in https://github.com/envoyproxy/envoy/blob/main/STYLE.md

File: source/extensions/filters/http/cache/cache_headers_utils.cc
Comment (jmarantz) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:311]: Yeah vary_identifier is being appended to in most cases. The only opportunity I see to improve this is to avoid the alloc in the case below where absl::nullopt is being returned.

You could do a pass over all the headers first and if any of them are not allowed return nullopt before trying to accumulate any strings. Not sure if that's worth it.

**Comment by capoferro** [comment]
Hm. Looks like coverage reporting is timing out consistently. Should we keep retesting, or just accept it as it is? It passed before the variable rename.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
